components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt

```diff
@@ -347,6 +359,66 @@ class GeckoEngineSession(
         }
     }
 
+    @Suppress("ComplexMethod")
+    internal fun createHistoryDelegate() = object : GeckoSession.HistoryDelegate {
+        @Suppress("TooGenericExceptionCaught")
+        override fun onVisited(
+            session: GeckoSession,
+            url: String,
+            lastVisitedURL: String?,
+            flags: Int
+        ): GeckoResult<Boolean>? {
+            if ((flags and GeckoSession.HistoryDelegate.VISIT_TOP_LEVEL) == 0 ||
+                (flags and GeckoSession.HistoryDelegate.VISIT_UNRECOVERABLE_ERROR) != 0) {
+
+                return GeckoResult.fromValue(false)
+            }
+
+            val isReload = lastVisitedURL?.let { it == url } ?: false
+            val result = GeckoResult<Boolean>()
+            scope.launch {
+                var exception: Throwable? = null
+                try {
+                    settings.historyTrackingDelegate?.onVisited(url, isReload, privateMode)
+                } catch (t: Throwable) {
+                    exception = t
+                }
+
+                if (exception == null) {
+                    // Don't mark links as visited in private mode.
+                    val markAsVisited = !privateMode
+                    result.complete(markAsVisited)
+                } else {
+                    result.completeExceptionally(exception)
+                }
+            }
+            return result
+        }
+
+        @Suppress("TooGenericExceptionCaught")
+        override fun getVisited(
+            session: GeckoSession,
+            urls: Array<out String>
+        ): GeckoResult<BooleanArray>? {
+            val result = GeckoResult<BooleanArray>()
+            scope.launch {
+                var exception: Throwable? = null
+                val visits: List<Boolean>? = try {
+                    settings.historyTrackingDelegate?.getVisited(urls.toList(), privateMode)?.await()
+                } catch (t: Throwable) {
+                    exception = t
+                    null
```

Comment:
Remove this?

Reply:
I'm assigning to `val visits`, so I think I still need the `null`, but it's moot if I remove the catches.